core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -245,9 +253,21 @@ final class ShuffleBlockFetcherIterator(
       case FetchBlockInfo(blockId, size, mapIndex) => (blockId.toString, (size, mapIndex))
     }.toMap
     val remainingBlocks = new HashSet[String]() ++= infoMap.keys
+    val deferredBlocks = new ArrayBuffer[String]()
     val blockIds = req.blocks.map(_.blockId.toString)
     val address = req.address
 
+    @inline def enqueueDeferredFetchRequestIfNecessary(): Unit = {
+      if (remainingBlocks.isEmpty && deferredBlocks.nonEmpty) {
+        val blocks = deferredBlocks.map { blockId =>
+          val (size, mapIndex) = infoMap(blockId)
+          FetchBlockInfo(BlockId(blockId), size, mapIndex)
+        }
+        results.put(DeferFetchRequestResult(FetchRequest(address, blocks.toSeq)))
+        deferredBlocks.clear()
+      }
+    }
```

Comment:
I will let @otterc elaborate more; we had discussed whether we can minimize direct memory load in `fetchUpToMaxBytes` due to all 'expensive' deferred blocks being sent out (mostly) together (modulo constraints on `maxBlocksInFlightPerAddress`/`maxBytesInFlight`).

Reply:
I'm not sure about the "direct memory load" you mean here. I'll wait for @otterc's explanation.

Reply:
This is related to the conversation here:
https://github.com/apache/spark/pull/32287#discussion_r623425119
We were discussing simple ways to reduce the number of remote fetch requests after the OOM. One such thing could be that after the OOM, we just sent the out the requests that were deferred due to the OOM and not send any additional requests. 
I am not sure how effective is this going to be though. Since the in-flight limit remains the same, the next call to `fetchUpToMaxBytes` when it sends out non-deferred requests can cause new blocks to OOM.

Another simple way could be to modify `isRemoteBlockFetchable` such that after this iterator has seen an OOM, it will also check `bytesInFlight + fetchReqQueue.front.size` < freeDirectorMemory?

Reply:
So after #32287 (comment), I have changed the unset condition to `freeDirectorMemory > maxReqSizeShuffleToMem (200M by default)`, which I think is already very strict. So, it should avoid the issue you mentioned in  #32287 (comment).

> Another simple way could be to modify isRemoteBlockFetchable such that after this iterator has seen an OOM, it will also check bytesInFlight + fetchReqQueue.front.size < freeDirectorMemory?

Do you mean check bytesInFlight + fetchReqQueue.front.size < freeDirectorMemory for all the cases or only when `isNettyOOMOnShuffle=true`?  If you also want to check when `isNettyOOMOnShuffle=false`, I'd like to mention that block size is not equal to the consumed memory size of Netty. As you know,  blocks that bigger than  `maxReqSizeShuffleToMem` would be stored on disk.

Reply:
> I have changed the unset condition to freeDirectorMemory > maxReqSizeShuffleToMem (200M by default), which I think is already very strict. So, it should avoid the issue you mentioned in #32287 (comment).

I don't think it will avoid the issue. This adds more time when the next set of requests are going to be set. However, the next set of requests (including the deferred ones) will still be sent at same frequency, so some of them would again see OOMs. 

> Do you mean check bytesInFlight + fetchReqQueue.front.size < freeDirectorMemory for all the cases or only when isNettyOOMOnShuffle=true? 

I meant that if once this OOM is encountered, after that the iterator checks against freeDirectorMemory as well. If a request.size > `maxReqSizeShuffleToMem` then we can skip that one as that is stored on disk.

Reply:
I think there are two different cases here. For example, assuming there're 100 requests.

For case a), the OOM threshold might be 80 requests. In this case, after OOMed, the deferred 20 requests shouldn't hit the OOM again. 

For case b), the OOM threshold might be 20 requests. In this case, there're still 80 deferred requests, which would hit the OOM soon as you mentioned. That being said, I think the current fix would work around the issue in the end. Note that the application would fail before the fix.

To improve the current fix further, I think we can do it in a separate PR as I think it's not an easy thing[1] to do (or do you have an idea now?) and must require more discussion. WDTY?

1. Even if we skip the case of request > maxReqSizeShuffleToMem, note that the in-memory request size is not strictly equal to the consumed memory size in Netty due to Netty's memory management mechanism (https://www.facebook.com/notes/facebook-engineering/scalable-memory-allocation-using-jemalloc/480222803919). For example, Netty may allocate 16MB for a 9MB block.
And there could be multiple tasks fetching concurrently. So we may need to track the total `bytesInFlight` of all tasks rather than the single task itself. And it would require more synchronization among tasks and make the thing more complex.

Reply:
It sounds like a good idea to dynamically tune the request frequency, but this doesn't seem like trivial work. I'd vote for doing it in a follow-up with more discussions.

Reply:
If nontrivial, we can definitely push it to a later effort - can you file a jira for it @Ngone51 so that we can track it (and perhaps someone can pick it up later ?).
This PR is already a marked improvement over what we currently have w.r.t OOM at executor - as long as ESS load does not go up :-)

Reply:
Filed the JIRA ticket: https://issues.apache.org/jira/browse/SPARK-35330